pretrain/tinyllama.py

```diff
@@ -80,12 +84,12 @@ def setup(resume: Union[bool, Path] = False):
     else:
         strategy = "auto"
 
-    fabric = L.Fabric(devices=devices, strategy=strategy, precision="bf16-mixed", loggers=[logger])
+    fabric = L.Fabric(devices=devices, strategy=strategy, precision="bf16-mixed", loggers=[_logger])
     fabric.launch()
 
     fabric.print(hparams)
-    if use_wandb:
-        logger.log_hyperparams(hparams)
+    if logger in ("tensorboard", "wandb"):
```

Comment:
if you used `isinstance` here, you could overwrite `logger: str` with `logger: Logger` when you create the instance and avoid the ugly underscore. this is fine because we don't run mypy

Or you could have the string variable be called `logger_kind` or `logger_type`

Reply:
I didn't do it because of mypy.
Can't write and read from the same name if one is a local and the other is a non local variable. I really don't want to do this.

Reply:
Maybe instead of 
```python
logger = "tensorboard"
```
use
```python
logs_provider = "tensorboard"
```
and then
```python
if logs_provider == "csv":
    logger = CSVLogger(...)
```
or maybe a different name, a better one ...

-----

> Can't write and read from the same name if one is a local and the other is a non local variable.

A dummy question: what about `nonlocal` keyword?

Reply:
@Andrei-Aksionov Thanks. I trying hard to keep this as boilerplate-free as possible, meaning no constructs like nonlocal expressions etc. Once I'm done with my pretraining efforts, I will probably remove the wandb logger again. So for now, I opted for pushing all this into a function and hope nobody complains.